dagobah/core/job.py

```diff
@@ -467,12 +469,119 @@ def initialize_snapshot(self):
         if not is_valid:
             raise DagobahError(reason)
 
-        self.snapshot = snapshot_to_validate
+        verified = self.verify()
+
+        if not verified:
+            raise DagobahError("Job has a cycle, possibly within another Job "
+                               + "reference")
+
+        # Due to thread locks in underlying tasks, they must be manually copied
+        tasks_copy = {}
+        for name, task in self.tasks.iteritems():
+            if self.implements_runnable(task):
+                tasks_copy[name] = Task(self, task.command, task.name,
+                                        soft_timeout=task.soft_timeout,
+                                        hard_timeout=task.hard_timeout,
+                                        hostname=task.hostname)
+            elif self.implements_expandable(task):
+                tasks_copy[name] = JobTask(self, task.name,
+                                           task.target_job_name)
+            else:
+                raise DagobahError("Malformed task neither a Task nor JobTask")
```

Comment:
If we end up needing to do this (which I really hope we don't), we should implement this sort of pattern as instance methods on the respective classes so they're easier to maintain and we don't end up duplicating that code anywhere else. This pattern is brittle as-is.